File: sdk/kotlin/src/main/kotlin/com/telecom/CurrencyAPI.kt
package com.telecom

import kotlinx.serialization.Serializable

/**
 * Currency and Billing API
 */
class CurrencyAPI(private val httpClient: HTTPClient) {
    
    /**
     * Convert currency
     */
    suspend fun convert(from: String, to: String, amount: Double): ConvertResponse {
        return httpClient.post("/api/v1/currency/convert",
            mapOf(
                "from" to from,
                "to" to to,
                "amount" to amount
            ))
    }
    
    /**
     * Get exchange rate between currencies
     */
    suspend fun getExchangeRate(from: String, to: String): ExchangeRate {
        return httpClient.get("/api/v1/currency/exchange/$from/$to")
    }
    
    /**
     * Get exchange rate history
     */
    suspend fun getExchangeRateHistory(from: String, to: String, days: Int = 30): List<ExchangeRate> {
        return httpClient.get("/api/v1/currency/exchange/$from/$to/history",
            mapOf("days" to days))
    }
    
    /**
     * Get supported currencies
     */
    suspend fun getSupportedCurrencies(): List<Currency> {
        return httpClient.get("/api/v1/currency/currencies")
    }
    
    /**
     * Refresh exchange rates
     */
    suspend fun refreshExchangeRates(): JsonObject {
        return httpClient.post("/api/v1/currency/exchange/refresh", emptyMap())
    }
    
    /**
     * Process billing transaction
     */
    suspend fun processBilling(billingData: Map<String, Any>): BillingTransaction {
        return httpClient.post("/api/v1/currency/billing", billingData)
    }
    
    /**
     * Get billing history for a profile
     */
    suspend fun getBillingHistory(profileId: String, limit: Int = 50): List<BillingTransaction> {
        return httpClient.get("/api/v1/currency/billing/history/$profileId",
            mapOf("limit" to limit))
    }
    
    /**
     * Get billing summary for a profile
     */
    suspend fun getBillingSummary(profileId: String, period: String = "monthly"): BillingSummary {
        return httpClient.get("/api/v1/currency/billing/summary/$profileId",
            mapOf("period" to period))
    }
    
    /**
     * Process refund
     */
    suspend fun processRefund(transactionId: String, reason: String): BillingTransaction {
        return httpClient.post("/api/v1/currency/billing/refund/$transactionId",
            mapOf("reason" to reason))
    }
    
    /**
     * Get billing analytics
     */
    suspend fun getBillingAnalytics(period: String = "monthly"): JsonObject {
        return httpClient.get("/api/v1/currency/billing/analytics",
            mapOf("period" to period))
    }
}

@Serializable
data class ConvertRequest(
    val from: String,
    val to: String,
    val amount: Double
)

@Serializable
data class ConvertResponse(
    val from: String,
    val to: String,
    val amount: Double,
    val converted: Double,
    val rate: Double,
    val timestamp: String
)

@Serializable
data class ExchangeRate(
    val from: String,
    val to: String,
    val rate: Double,
    val timestamp: String
)

@Serializable
data class Currency(
    val code: String,
    val name: String,
    val symbol: String
)

@Serializable
data class BillingTransaction(
    val id: String,
    val profileId: String,
    val amount: Double,
    val currency: String,
    val type: String,
    val status: String,
    val description: String,
    val createdAt: String
)

@Serializable
data class BillingSummary(
    val profileId: String,
    val period: String,
    val totalAmount: Double,
    val currency: String,
    val transactionCount: Int,
    val breakdown: Map<String, Double>
)

File: sdk/swift/Sources/TelecomSDK/CurrencyAPI.swift
import Foundation

/// Currency and Billing API
public class CurrencyAPI {
    private let client: HTTPClient
    
    public init(client: HTTPClient) {
        self.client = client
    }
    
    // MARK: - Currency Conversion
    
    /// Convert currency
    public func convert(from: String, to: String, amount: Double) async throws -> ConvertResponse {
        return try await client.post(
            "/api/v1/currency/convert",
            body: [
                "from": from,
                "to": to,
                "amount": amount
            ],
            responseType: ConvertResponse.self
        )
    }
    
    /// Get exchange rate between currencies
    public func getExchangeRate(from: String, to: String) async throws -> ExchangeRate {
        return try await client.get(
            "/api/v1/currency/exchange/\(from)/\(to)",
            responseType: ExchangeRate.self
        )
    }
    
    /// Get exchange rate history
    public func getExchangeRateHistory(from: String, to: String, days: Int = 30) async throws -> [ExchangeRate] {
        return try await client.get(
            "/api/v1/currency/exchange/\(from)/\(to)/history",
            parameters: ["days": days],
            responseType: [ExchangeRate].self
        )
    }
    
    /// Get supported currencies
    public func getSupportedCurrencies() async throws -> [Currency] {
        return try await client.get(
            "/api/v1/currency/currencies",
            responseType: [Currency].self
        )
    }
    
    /// Refresh exchange rates
    public func refreshExchangeRates() async throws -> [String: Any] {
        return try await client.post("/api/v1/currency/exchange/refresh", body: [:])
    }
    
    // MARK: - Billing
    
    /// Process billing transaction
    public func processBilling(_ billingData: [String: Any]) async throws -> BillingTransaction {
        return try await client.post(
            "/api/v1/currency/billing",
            body: billingData,
            responseType: BillingTransaction.self
        )
    }
    
    /// Get billing history for a profile
    public func getBillingHistory(profileId: String, limit: Int = 50) async throws -> [BillingTransaction] {
        return try await client.get(
            "/api/v1/currency/billing/history/\(profileId)",
            parameters: ["limit": limit],
            responseType: [BillingTransaction].self
        )
    }
    
    /// Get billing summary for a profile
    public func getBillingSummary(profileId: String, period: String = "monthly") async throws -> BillingSummary {
        return try await client.get(
            "/api/v1/currency/billing/summary/\(profileId)",
            parameters: ["period": period],
            responseType: BillingSummary.self
        )
    }
    
    /// Process refund
    public func processRefund(transactionId: String, reason: String) async throws -> BillingTransaction {
        return try await client.post(
            "/api/v1/currency/billing/refund/\(transactionId)",
            body: ["reason": reason],
            responseType: BillingTransaction.self
        )
    }
    
    /// Get billing analytics
    public func getBillingAnalytics(period: String = "monthly") async throws -> [String: Any] {
        return try await client.get(
            "/api/v1/currency/billing/analytics",
            parameters: ["period": period]
        )
    }
}

// MARK: - Currency Types

public struct ConvertRequest: Codable {
    public let from: String
    public let to: String
    public let amount: Double
}

public struct ConvertResponse: Codable {
    public let from: String
    public let to: String
    public let amount: Double
    public let converted: Double
    public let rate: Double
    public let timestamp: String
}

public struct ExchangeRate: Codable {
    public let from: String
    public let to: String
    public let rate: Double
    public let timestamp: String
}

public struct Currency: Codable {
    public let code: String
    public let name: String
    public let symbol: String
}

public struct BillingTransaction: Codable {
    public let id: String
    public let profileId: String
    public let amount: Double
    public let currency: String
    public let type: String
    public let status: String
    public let description: String
    public let createdAt: String
}

public struct BillingSummary: Codable {
    public let profileId: String
    public let period: String
    public let totalAmount: Double
    public let currency: String
    public let transactionCount: Int
    public let breakdown: [String: Double]
}
